### PR TITLE
builtins: introduce hlc_to_timestamp

### PR DESCRIPTION
**Pull request by otan**
There is a crdb_internal.approximate_timestamp, which does something
similar but returns a timestamp and is "internal". This is a public
facing version which I think we should keep, returning TimestampTZ so it
returns nicely if the user decides to set the timezone.

Release note (sql change): Introduce a hlc_to_timestamp builtin, which
converts a CockroachDB HLC to a TimestampTZ. This is useful for pretty
printing crdb_internal_mvcc_timestamp or cluster_logical_timestamp(),
but is not useful for accuracy.

### CODE REVIEW

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -3695,19 +3695,41 @@ func TimestampToDecimal(ts hlc.Timestamp) apd.Decimal {
...
+// DecimalToInexactDTimestamp is the inverse of TimestampToDecimal. It converts
+// a decimal constructed from an hlc.Timestamp into an approximate DTimestamp
+// containing the walltime of the hlc.Timestamp.
+func DecimalToInexactDTimestamp(d *DDecimal) (*DTimestamp, error) {
+	ts, err := decimalToHLC(d)
+	if err != nil {
+		return nil, err
 	}
-	return TimestampToInexactDTimestamp(hlc.Timestamp{WallTime: coef.Int64()}), nil
+	return TimestampToInexactDTimestamp(ts), nil
```
Comment (rafiss) [inline pkg/sql/sem/tree/eval.go:3732]: it seems like we could do away with the `TimestampToInexactDTimestamp` function. so this line would change to
```
return MustMakeDTimestamp(timeutil.Unix(0, ts.WallTime), time.Microsecond), nil
```

(just to make the impls of `DecimalToInexactDTimestampTZ` and `DecimalToInexactDTimestamp` more similar to each other to increase readability.)

Comment (otan) [reply pkg/sql/sem/tree/eval.go:3732]: `TimestampToInexactDTimestamp` is used in a couple of places in crdb_internal.go, rather not replicate the logic

**Review by nvb — APPROVED** [review]
:lgtm: once Rafi signs off.

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68360#-:-MgGXWeq-4LLu8rApJUb:bz0tzbb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/sem/builtins/builtins.go, line 2572 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68360#-MgGX5I53caJpSy01R8E:-MgGX5I53caJpSy01R8F:b-a0p0oq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5822e4482149e50034cb8364fcb931e8e89b97c6/pkg/sql/sem/builtins/builtins.go#L2572)):*
> ```Go
> 				return tree.DecimalToInexactDTimestampTZ(&d)
> 			},
> 			Info: `Returns a TimestampTZ representation of a CockroachDB HLC in decimal form.
> ```

Should we update the docs on `cluster_logical_timestamp` to reference "CockroachDB HLC in decimal form" as well?

File: pkg/sql/sem/tree/casts.go [hunk 2]
```diff
@@ -178,7 +178,7 @@ var validCasts = []castInfo{
 	{from: types.TimestampFamily, to: types.StringFamily, volatility: VolatilityImmutable},
 	{
 		from: types.TimestampTZFamily, to: types.StringFamily, volatility: VolatilityStable,
-		volatilityHint: "TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC')::STRING instead.",
+		volatilityHint: "TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC') instead.",
```
Comment (rafiss) [inline pkg/sql/sem/tree/casts.go:181]: looks like this change snuck in, but it's an OK docs fix

**Review by rafiss — APPROVED** [review]

File: pkg/sql/sem/tree/casts.go (hunk 2 above)
Comment (otan) [reply pkg/sql/sem/tree/casts.go:181]: i was wondering where this went.

**Comment by otan** [comment]
bors r=rafiss,nvanbeschoten

**Pull request merged**